### PR TITLE
[ENH] Hidden layer droput uniformization in the deep learning models. #9103

### LINKED ISSUE #9103: [ENH] Hidden layer droput uniformization in the deep learning models.
### Is your feature request related to a problem? Please describe.
Hidden layer droput uniformization in the deep learning models (classifiers, regressors & networks).

Current behaviour:
- In some of the TensorFlow based deep learning models, hidden layer dropout is hardcoded and not exposed to the users. I have pasted one such example below for illustration.
https://github.com/sktime/sktime/blob/071551b59669a946a386617fb7706e69349e5dff/sktime/networks/rnn/_rnn_tf.py#L123
- In Deep Learning models, currently we can only set the hidden layer activations globally but there is no support for setting it layer-wise. Or, if it is set layer-wise then it is hardcoded and not exposed to the end-user. For example, in `MLPNetwork` the dropout is different in each layer because the implementation is based on the model described in section `3.2.1 Multi Layer Perceptron` on page 14 in [Deep learning for time series classification: a review](https://arxiv.org/pdf/1809.04356) paper from 2019
https://github.com/sktime/sktime/blob/071551b59669a946a386617fb7706e69349e5dff/sktime/networks/mlp.py#L68-L77

---

### Describe the solution you'd like

Expected/Suggested behaviour:
1. Wherever dropouts are used in the hidden layers, they should be exposed to the end-user via `dropout` parameter.
2. Currently, parameter `dropout` (wherever present) is expected to be a `float`, change it to expect either a `tuple` of floats or a single `float`. And use the tuple of floats to set layer-wise dropouts, whenever it is specified.
3. Care needs to be taken in handling the dropouts specified via a tuple, such as enforcing the length of the passed tuple be equal to the number of hidden layers specified, and using the correct dropout in each corresponding hidden layer, etc.
4. Care needs to be taken where it is allowed and not-allowed to set the dropout layer-wise. There may be certain network implementations either in TensorFlow or PyTorch which do not allow setting layer-wise dropouts in the network, in such cases parameter `dropout` should strictly be a `float` only and DOCSTRING should reflect the same.
5. In current behaviour section above, I have mentioned only 2 examples but all deep learning models should be investigated and fix should be implemented everywhere it is needed.
6. In order to maintain the consistent behaviour, while setting the default value of `dropout` parameter 
    1. use the same value/s which are currently hard-coded,
    2. if it is not hard-coded then use the default value from the underlying library's implementation/documentation.

---

### Additional context
This issue originated from the discussion in the [comments](https://github.com/sktime/sktime/pull/9042#discussion_r2490855188) of #9042

### PR DESCRIPTION
**Pull request by abhimanyudalal1**
Fixes #9103

#### Reference Issues/PRs

This PR exposes the `dropout` parameter in `MLPNetwork` and `RNNNetwork`. This allows users to control dropout regularization in deep learning models while preserving backward compatibility.
### Changes Made
1.  **MLPNetwork (`mlp.py`):**
    * Added `dropout` parameter to `__init__`.
    * Default: `(0.1, 0.2, 0.2, 0.3)` (Tuple) to match previous hardcoded values.
    * Logic: Supports `float` (uniform dropout) or `tuple` (layer-wise dropout). Validates that tuple length matches network layers.

2.  **RNNNetwork (`_rnn_tf.py`):**
    * Added `dropout` parameter to `__init__`.
    * Default: `0.0` (Float) to match previous hardcoded behavior.
    * Logic: Passed directly to the `keras.layers.SimpleRNN` layer.

* Existing code relying on default behavior will produce identical networks.

#### What should a reviewer concentrate their feedback on?
* Check the if/elif/else block in MLPNetwork (lines 63-77) that handles the float/tuple logic. I plan to replicate this pattern in other networks (cntc, lstmfcn, gru) if this approach is approved.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

File: sktime/classification/deep_learning/rnn/_rnn_tf.py [hunk 1]
```diff
@@ -55,6 +55,8 @@ class SimpleRNNClassifier(BaseDeepClassifier):
         whether the layer uses a bias vector.
     optimizer : keras.optimizers object, default = RMSprop(lr=0.001)
         specify the optimizer and the learning rate to be used.
+        dropout : float, default=0.0
+    The dropout rate for the RNN layer.
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/rnn/_rnn_tf.py]: Formatting of the docstring is non-conforming to the style. 
Fix: De-tab the line that mentions the name of the parameter and tab the line that describes functionality of the parameter.

File: sktime/networks/mlp.py [hunk 2]
```diff
@@ -59,21 +60,37 @@ def build_network(self, input_shape, **kwargs):
         input_layer : a keras layer
         output_layer : a keras layer
         """
+        if isinstance(self.dropout, (float, int)):
```
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: The docstring mentions the dropout to be either `float` or `tuple`. 
Strictly speaking, `int` shouldn't be allowed here. 
Perhaps you're thinking of handling `0` or `1`, but even in those cases `0.0` is the standard practice and `1.0` is mostly an invalid setting as it leads to 100% dropout.

File: sktime/classification/deep_learning/mlp.py [hunk 3]
```diff
@@ -44,6 +44,9 @@ class MLPClassifier(BaseDeepClassifier):
         whether the layer uses a bias vector.
     optimizer : keras.optimizers object, default = Adam(lr=0.01)
         specify the optimizer and the learning rate to be used.
+    dropout : float or tuple, default=(0.1, 0.2, 0.2, 0.3)
+        Dropout rates for hidden layers. If float, same rate for all layers.
+        If tuple, must have length 4 for each layer.
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/mlp.py]: May I suggest rewording this line to something like: If tuple, it must have length equal to number of hidden layers in the MLP, each element specifying the dropout rate for the corresponding hidden layer. Current implementation of the MLP has 4 hidden layers.

It is helpful to end-user if we explain the reasoning behind the length.

File: sktime/networks/mlp.py [hunk 4]
```diff
@@ -20,6 +20,10 @@ class MLPNetwork(BaseDeepNetwork):
         activation function used for hidden layers;
         List of available keras activation functions:
         https://keras.io/api/layers/activations/
+    dropout : float or tuple, default=(0.1, 0.2, 0.2, 0.3)
+        The dropout rate for the hidden layers.
+        If float, the same rate is used for all layers.
+        If tuple, must be of length 4, specifying the rate for each layer.
```
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: Same comment as the MLP classifier docstring.

File: sktime/regression/deep_learning/mlp.py [hunk 5]
```diff
@@ -43,6 +43,9 @@ class MLPRegressor(BaseDeepRegressor):
         whether the layer uses a bias vector.
     optimizer : keras.optimizers object, default = Adam(lr=0.01)
         specify the optimizer and the learning rate to be used.
+    dropout : float or tuple, default=(0.1, 0.2, 0.2, 0.3)
+        Dropout rates for hidden layers. If float, same rate for all layers.
+        If tuple, must have length 4 for each layer.
```
Comment (RecreationalMath) [inline sktime/regression/deep_learning/mlp.py]: Same comment as the MLP classifier docstring.

**Comment by abhimanyudalal1** [comment]
I have now changed the files using MLP and RNN networks specifically:

MLPClassifier / MLPRegressor &

RNNClassifier / RNNRegressor

File: sktime/networks/mlp.py (hunk 2 above)
Comment (abhimanyudalal1) [reply sktime/networks/mlp.py]: The thought was even if the user enters 0 (int, probably the only case where int is entered), it would still be used as 0.0 according to the next line-
  dropout_rates = [float(self.dropout)] * 4
But yeah, as standard practice, I'll remove int dtype.

File: sktime/networks/mlp.py [hunk 6]
```diff
@@ -59,21 +62,37 @@ def build_network(self, input_shape, **kwargs):
         input_layer : a keras layer
         output_layer : a keras layer
         """
+        if isinstance(self.dropout, (float)):
```
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: Is there a reason for using parenthesis around `float`?

File: sktime/networks/mlp.py [hunk 7]
```diff
@@ -59,21 +62,37 @@ def build_network(self, input_shape, **kwargs):
...
+            dropout_rates = [float(self.dropout)] * 4
```
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: This statement is inside the dropout's float-type-check `if` block and will only be reached if `dropout` is indeed a float. Typecasting it again to a float is redundant and can be removed.

File: sktime/networks/mlp.py [hunk 8]
```diff
@@ -59,21 +62,37 @@ def build_network(self, input_shape, **kwargs):
...
+        elif isinstance(self.dropout, tuple):
+            if len(self.dropout) != 4:
+                raise ValueError(
+                    "If `dropout` is a tuple, it must be of length 4"
```
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: As you did in the docstring, please mention the reasoning here as well - If `dropout` is a tuple, it must have length equal to number of hidden layers in the MLP, each element specifying the dropout rate for the corresponding hidden layer. Current implementation of the MLP has 4 hidden layers, hence the tuple must be of length 4.

These error messages are useful for the enduser and having them descriptive makes for a better traceback when enduser runs into an error.

File: sktime/networks/mlp.py [hunk 9]
```diff
@@ -59,21 +62,37 @@ def build_network(self, input_shape, **kwargs):
...
+                    "for MLPNetwork. "
+                    f"Found length of {len(self.dropout)}"
```
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: Minor change: Found length of `dropout` to be: {len(self.dropout)}

File: sktime/networks/mlp.py (hunk 7 above)
Comment (RecreationalMath) [inline sktime/networks/mlp.py]: `sktime` has a standard way of naming helper/intermittent variables, I will try to find the link to the guide and paste here. 
Essentially, it is recommended to put a `_` in front of the existing variable name and use that.
So in this case, your helper variable would be `_dropout` instead of `dropout_rates`.

File: sktime/networks/mlp.py (hunk 6 above)
Comment (abhimanyudalal1) [reply sktime/networks/mlp.py]: Not anymore, earlier it was for giving an option for int dtype. Will remove.

File: sktime/networks/mlp.py (hunk 7 above)
Comment (abhimanyudalal1) [reply sktime/networks/mlp.py]: Okay, I've made the changes. A link to the guide you mentioned would be really helpful, thanks.

Comment (RecreationalMath) [reply sktime/networks/mlp.py]: @fkiraly I can't locate the documentation page which specifies private and public variable naming conventions used in sktime i.e. usage of underscore before the name of private variables. Could you please point to it here? Thanks!

File: sktime/networks/mlp.py (hunk 2 above)
Comment (RecreationalMath) [reply sktime/networks/mlp.py]: > hey, I included int in the check to safely handle the case for dropout=0. Since isinstance(0, float) returns False in Python, removing int from the check would cause the code to raise a TypeError for integer inputs, which is bad UX.
are you sure we should give only float input option?

My way of confirming this is to follow the documentation and see if having `int` as the type of the `p` parameter is permissive or not. [PyTorch docs for dropout](https://docs.pytorch.org/docs/stable/generated/torch.nn.Dropout.html) suggest only floats to be used and if you look at the actual implementation in [lines 18-20 in pytorch/torch/nn/modules/dropout.py in the official torch repo](https://github.com/pytorch/pytorch/blob/d38164a545b4a4e4e0cf73ce67173f70574890b6/torch/nn/modules/dropout.py#L18-L20) it expects the type of parameter `p` as `float`.

This is my thought process on this. Other devs ( @phoeenniixx  / @fkiraly  ?) can confirm if allowing `int` is helpful and should be allowed.

File: sktime/networks/mlp.py (hunk 7 above)
Comment (fkiraly) [reply sktime/networks/mlp.py]: there is no such document, the conventions are as in general python, with some small additional, not strictly enforced ones:
* class/object variables that are not `__init__` args should usually be private (underscore start)
* fitted variables, assigned in `fit`, should have an underscore at the end (this convention comes from scikit-learn) - these get automatically exposed via `get_fitted_params`
